Python/optimizer.c

```diff
@@ -698,7 +698,7 @@ uop_optimize(
         return trace_length;
     }
     OBJECT_STAT_INC(optimization_traces_created);
-    _PyUOpExecutorObject *executor = PyObject_New(_PyUOpExecutorObject, &UOpExecutor_Type);
+    _PyUOpExecutorObject *executor = PyObject_NewVar(_PyUOpExecutorObject, &UOpExecutor_Type, trace_length+1);
```

Comment:
The ```trace_length+1``` value is used as line 708 adds a sentinel uop, causing the number of instructions to be ```trace_length+1```.

Reply:
Could you please add a comment here so that we don't forget, or to help others understand?

Reply:
Then again, I think the allocator will allocate an extra item anyways because of
```
.tp_basicsize = sizeof(_PyUOpExecutorObject),
```
If you compare to `PyTuple_Type`, it uses this:
```
sizeof(PyTupleObject) - sizeof(PyObject *),
```
so that the basic size excludes the size of the dummy array in the `PyTupleObject` type:
```
    PyObject *ob_item[1];
```
I do like that we explicitly have `+1` here to make it clear that we intend to use the sentinel value.

However, now that the executor is variable-length and *always* has room for the sentinel, I think that on line 708 below you need to set the sentinel unconditionally, i.e. remove the condition on line 707.

Please also update `uop_len` to just return `ob_size` instead of counting until the sentinel. (This also means we could just get rid of the sentinel -- I think it only exists for `uop_size`, and it's never reachable by the interpreter.)